bot/inLineButtons_quest.py

```diff
@@ -58,31 +89,62 @@ async def ask_question(message: types.Message, state: FSMContext, state_name: St
     for option in question["options"]:
         markup.row(KeyboardButton(text=option))
 
-    await message.answer(question["question"], reply_markup=markup.as_markup(resize_keyboard=True, one_time_keyboard=True))
+    await message.answer(question["question"], reply_markup=markup.as_markup(resize_keyboard=True, one_time_keyboard=False))
     await state.set_state(state_name)
 
+async def handle_answer(message: types.Message, state: FSMContext, next_state: State, question_index: int):
+    user_input = message.text
+    correct_options = questions[question_index]["options"]
+    if user_input not in correct_options:
+        print(f"Некорректный ответ от пользователя {message.from_user.id}: {user_input}.")
+        await message.answer("Пожалуйста, выберите один из предложенных вариантов ответа.")
+        return
+    user_answers[message.from_user.id].append(user_input)
+    await ask_question(message, state, next_state, question_index + 1)
 @dp.message(StateFilter(TestStates.Q1))
 async def process_q1(message: types.Message, state: FSMContext):
-    user_answers[message.from_user.id].append(message.text)
-    await ask_question(message, state, TestStates.Q2, 1)
+    await handle_answer(message, state, TestStates.Q2, 0)
 
 @dp.message(StateFilter(TestStates.Q2))
 async def process_q2(message: types.Message, state: FSMContext):
-    user_answers[message.from_user.id].append(message.text)
-    await ask_question(message, state, TestStates.Q3, 2)
+    await handle_answer(message, state, TestStates.Q3, 1)
 
 @dp.message(StateFilter(TestStates.Q3))
 async def process_q3(message: types.Message, state: FSMContext):
-    user_answers[message.from_user.id].append(message.text)
-    await ask_question(message, state, TestStates.Q4, 3)
+    await handle_answer(message, state, TestStates.Q4, 2)
 
 @dp.message(StateFilter(TestStates.Q4))
 async def process_q4(message: types.Message, state: FSMContext):
+    await handle_answer(message, state, TestStates.Q5, 3)
+
+@dp.message(StateFilter(TestStates.Q5))
+async def process_q5(message: types.Message, state: FSMContext):
+    await handle_answer(message, state, TestStates.Q6, 4)
+
+@dp.message(StateFilter(TestStates.Q6))
+async def process_q6(message: types.Message, state: FSMContext):
+    await handle_answer(message, state, TestStates.Q7, 5)
+
+@dp.message(StateFilter(TestStates.Q7))
+async def process_q7(message: types.Message, state: FSMContext):
+    await handle_answer(message, state, TestStates.Q8, 6)
+
+@dp.message(StateFilter(TestStates.Q8))
+async def process_q8(message: types.Message, state: FSMContext):
+    await handle_answer(message, state, TestStates.Q9, 7)
+
+@dp.message(StateFilter(TestStates.Q9))
+async def process_q9(message: types.Message, state: FSMContext):
+    await handle_answer(message, state, TestStates.Q10, 8)
+
+@dp.message(StateFilter(TestStates.Q10))
+async def process_q10(message: types.Message, state: FSMContext):
```

Comment:
Are there any options to avoid such repeat of the code? Or this is the only solution with states of questions?

Reply:
I tried to throw most of the code into the class. At the moment, I think this is the best solution.